docs/proposals/05-etcd-operator-tasks.md

```diff
@@ -0,0 +1,364 @@
+  // ObservedGeneration is the most recent generation observed for the resource.
+  ObservedGeneration *int64 `json:"observedGeneration,omitempty"`
+  // Status of the task
+  Status TaskState `json:"taskStatus"`
+  // Time at which operation has been triggered.
+  InitiatedAt metav1.Time `json:"initiatedAt"`
+  // LastError represents the errors when processing the task.
+  LastErrors []LastError `json:"lastErrors,omitempty"`
+  // Captures the last operation
+  // +optional
+  LastOperation *LastOperation `json:"lastOperation,omitempty"`
+  // Conditions represents the latest available observations of an object's current state.
+  // +optional
+  Conditions []metav1.Condition `json:"conditions,omitempty"`
+}
+
+type LastOperation struct {
+  // Name of the LastOperation.
+  Name opsName `json:"name"`
+  // Status of the last operation, one of pending, progress, completed, failed.
+  State OperationState `json:"state"`
+  // Last time the operation state transitioned from one to another.
+  LastTransitionTime metav1.Time `json:"lastTransitionTime"`
+  // A human readable message indicating details about the last operation.
+  Reason string `json:"reason"`
+}
+
+```
+
+### Lifecycle
+
+#### Creation
+
+Task(s) can be created by creating an instance of the `EtcdOperatorTask` custom resource specific to a task.
+
+> Note: In future, either a `kubectl` extension plugin or a `druidctl` will be introduced. Dedicated sub-commands will be created for each `out-of-band` task. This will drastically increase the usability for an operator for such tasks as the CLI extension will automatically created relevant instance of `EtcdOperatorTask` with the provided configuration.
+
+#### Execution
+
+* Authors propose to introduce a new controller (let's call it `operator-task-controller`) which watches for `EtcdOperatorTask` custom resource specific to a task defined by [.spec.taskType](#spec).
+* Each `out-of-band` task may have some task specific configuration defined in [.spec.taskConfig](#spec).
+* The controller (`operator-task-controller`) needs to parse this task specific config, which comes as a RawExtension, according to the schema defined for each task.
+* Moreover, all tasks have to adhere to some prerequisites(a.k.a `pre-conditions`) which will be necessary to execute the task. Authors propose to define pre-conditions for each task, which must be met for the task to be eligible for execution otherwise that task should be rejected.
+* If multiple tasks are invoked simultaneously or in `pending` state, then they will be executed in a First-In-First-Out (FIFO) manner.
+
+> Note: Dependent ordering among tasks will be addresed later which may enable concurrent execution of tasks.
+
+#### Deletion
+
+`Etcd-druid` will ensure that it garbage collects the custom resources and any other k8s resources created to realise the task as per `.spec.ttlSecondsAfterFinished` if defined in the [spec](#spec) or it assumes a default expiry time.
+
+### Use Cases
```

Comment:
I was wondering looking at the below use case and the possibility of integrating esp. with `gardenctl` can we have options where an operator might like to run an operation across shoots for e.g. 
- some of the selected shoots where he see a need 
- all shoots in a project as the problem is faced by all of them due to some nuance in the behavior 
- all shoots in a seed as we need to do some remediation across all shoots 
Does this change the design or implementation from a druid perspective esp. like selectors for which etcd to pick. 

Just thinking out loud, so feel free to ignore.

Reply:
You can do this now as well by creating one operator task per etcd resource. 
Regarding integration with `gardenctl` - yes it can be done for g/g use cases but we are now positioning druid to be used outside g/g as well and therefore what you ask can easily be one operation for `druidctl`. In this proposal we have not detailed that aspect as that will be a separate proposal/document detailing out what is possible using druid specific ctl extension.

Reply:
>You can do this now as well by creating one operator task per etcd resource.
Yes, but this response is good for feature tick point of view, consumption wise i'll go with the label selector approach you mentioned. 

Are you planning to have it in scope then? Its not obvious from the docs that you have such an intent.

Reply:
>Are you planning to have it in scope then? Its not obvious from the docs that you have such an intent.

This will be part of the CLI and that is not part of this proposal.